### PR TITLE
Fix error on importing `Events` in python3.11

### LINKED ISSUE #2904: Cannot import ignite.distributed on Python3.11
## 🐛 Bug description

Importing `ignite.distributed` fails on Python3.11. 

To reproduce: 
```bash
python3.11 -m pip install pytorch-ignite
python3.11 -c 'import ignite.distributed'
```

I get the following `AttributeError`: 
```python
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/__init__.py", line 3, in <module>
    import ignite.engine
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/engine/__init__.py", line 7, in <module>
    from ignite.engine.deterministic import DeterministicEngine
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/engine/deterministic.py", line 11, in <module>
    from ignite.engine.engine import Engine
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/engine/engine.py", line 13, in <module>
    from ignite.engine.events import CallableEventWithFilter, EventEnum, Events, EventsList, RemovableEventHandle, State
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/engine/events.py", line 254, in <module>
    class Events(EventEnum):
  File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/enum.py", line 560, in __new__
    raise exc
  File "/opt/homebrew/Cellar/python@3.11/3.11.2_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/enum.py", line 280, in __set_name__
    enum_member = enum_class._value2member_map_[value]
                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/ignite/engine/events.py", line 200, in __hash__
    return hash(self._name_)
                ^^^^^^^^^^^
AttributeError: 'CallableEventWithFilter' object has no attribute '_name_'. Did you mean: 'name'?
```

## Environment

- PyTorch Version: 2.0.0
- Ignite Version: 0.4.11
- OS (e.g., Linux): macOS Ventura 13.2.1
- How you installed Ignite: `pip`
- Python version: 3.11

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2904

### CODE REVIEW

File: ignite/engine/events.py [hunk 1]
```diff
@@ -237,7 +237,13 @@ def function_before_backprop(engine):
                 # ...
     """
 
-    pass
+    def __new__(cls, value: str) -> "EventEnum":
+        obj = CallableEventWithFilter.__new__(cls)
+        obj._value_ = value
+        return obj
+
+    def __repr__(self) -> str:
```
Comment (vfdev-5) [inline ignite/engine/events.py:245]: @sadra-barikbin is this really necesary ?

Comment (sadra-barikbin) [reply ignite/engine/events.py]: `__repr__` wasn't necessary. I removed it. But `__new__` is.

**Comment by vfdev-5** [comment]
Thanks @sadra-barikbin !

From a discussion on discord: 
> As of python3.11's next release we could remove __new__ from EventEnum, https://github.com/python/cpython/issues/103479

File: ignite/engine/events.py (hunk 1 above)
Comment (sadra-barikbin) [reply ignite/engine/events.py]: @vfdev-5 , from #2997 turns out it was necessary. Shall I add it as a follow-up or we should ask 3.11 users to update to `3.11.4`?

**Pull request merged**